### PR TITLE
Add perfect forwarding push-like functions of resources

### PR DESCRIPTION
**Pull request by avinabadasgupta**

### CODE REVIEW

**Comment by avinabadasgupta** [comment]
Adding tests tomorrow.

**Comment by avinabadasgupta** [comment]
@klingaard , while moving ```SharedData<DataT>```, I was not able to move the template type ```DataT```, it was always getting copied. 
Found out ```SharedData<DataT>``` internally has an array of ```ValidValue<DataT>``` and ```ValidValue<DataT>``` only had a copy assign operator and no move assign operator.

**Comment by klingaard** [comment]
> @klingaard , while moving `SharedData<DataT>`, I was not able to move the template type `DataT`, it was always getting copied.
> Found out `SharedData<DataT>` internally has an array of `ValidValue<DataT>` and `ValidValue<DataT>` only had a copy assign operator and no move assign operator.

Yeah, I'm seeing odd behavior in general with `SharedData` -- we're using it in a fashion that requires it to be copied.  But, it's not behaving properly on copy (I think the `GlobalEvent` it contains is getting hosed).  Curious if you ran into anything like that when you played around with it.

**Comment by klingaard** [comment]
Curious, do you have tests that ensure that the data _isn't_ moved?  There are many use cases where a modeler will insert/push data into a `Queue`, `Array`, etc and then continue using the data immediately following the addition.

**Comment by avinabadasgupta** [comment]
> Curious, do you have tests that ensure that the data _isn't_ moved? There are many use cases where a modeler will insert/push data into a `Queue`, `Array`, etc and then continue using the data immediately following the addition.

Everything else, apart from the code snippets I put in the tests, use data copy. I will add some checks in the tests to ensure data isn't copied.

**Comment by avinabadasgupta** [comment]
> Curious, do you have tests that ensure that the data _isn't_ moved? There are many use cases where a modeler will insert/push data into a `Queue`, `Array`, etc and then continue using the data immediately following the addition.

Added tests to ensure copy works as expected and no invalidation of passed-in data happens.

**Comment by klingaard** [comment]
Thanks.  I'll look this over tomorrow morning

File: sparta/sparta/resources/Buffer.hpp [hunk 1]
```diff
@@ -895,6 +877,85 @@ namespace sparta
...
+            validator_->attachDataPointer(free_position_);
+            ++num_valid_;
+            free_position_ = free_position_->next_free;
+            updateUtilizationCounters_();
+
+            return entry;
+        }
+
+        template<typename U>
+        iterator insertImpl_(uint32_t idx, U&& dat)
```
Comment (klingaard) [inline sparta/sparta/resources/Buffer.hpp]: I wonder if this function can call the above one before pushing everything down...

**Comment by klingaard** [comment]
Before you merge this, let me try with our internal simulator to make sure there isn't a use case missing...

**Comment by klingaard** [comment]
> Before you merge this, let me try with our internal simulator to make sure there isn't a use case missing...

All regressions passed!  You're good to go!

**Comment by avinabadasgupta** [comment]
> > Before you merge this, let me try with our internal simulator to make sure there isn't a use case missing...
> 
> All regressions passed! You're good to go!

Merging it in.

**Pull request merged**